src/plugin.ts

```diff
@@ -0,0 +1,534 @@
+    driverAbortSignal?.addEventListener('abort', driverAbortCb);
+    const signal = combinedAbortController.signal;
+
+    try {
+      get_tokens: {
+        if ((state.currentTokenSet?.set?.expires_in ?? 0) > 5 * 60) {
+          this.logger.emit('mongodb-oidc-plugin:skip-auth-attempt', {
+            reason: 'not-expired',
+          });
+          break get_tokens;
+        }
+        if (await state.currentTokenSet?.tryRefresh?.()) {
+          this.logger.emit('mongodb-oidc-plugin:skip-auth-attempt', {
+            reason: 'refresh-succeeded',
+          });
+          break get_tokens;
+        }
+        state.currentTokenSet = null;
+        let error;
+        if (this.isFlowAllowed('auth-code')) {
+          try {
+            this.logger.emit('mongodb-oidc-plugin:auth-attempt-started', {
+              flow: 'auth-code',
+            });
+            await this.authorizationCodeFlow(state, signal);
+            this.logger.emit('mongodb-oidc-plugin:auth-attempt-succeeded');
+            break get_tokens;
+          } catch (err: unknown) {
+            error = err;
+            this.logger.emit('mongodb-oidc-plugin:auth-attempt-failed', {
+              error: errorString(err),
+            });
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            if (!(err as any)?.[kEnableFallback]) throw err;
+          }
+        }
+        if (this.isFlowAllowed('device-auth')) {
+          try {
+            this.logger.emit('mongodb-oidc-plugin:auth-attempt-started', {
+              flow: 'device-auth',
+            });
+            await this.deviceAuthorizationFlow(state, signal);
+            this.logger.emit('mongodb-oidc-plugin:auth-attempt-succeeded');
+            break get_tokens;
+          } catch (err: unknown) {
+            this.logger.emit('mongodb-oidc-plugin:auth-attempt-failed', {
+              error: errorString(err),
+            });
+            throw err;
+          }
+        }
+        if (error) throw error;
+      }
+
+      if (!state.currentTokenSet?.set?.access_token) {
+        throw new MongoDBOIDCError('Could not retrieve valid access token');
+      }
+    } catch (err: unknown) {
+      this.logger.emit('mongodb-oidc-plugin:auth-failed', {
+        error: errorString(err),
+      });
+      throw err;
+    } finally {
+      this.options.signal?.removeEventListener('abort', optionsAbortCb);
+      driverAbortSignal?.removeEventListener('abort', driverAbortCb);
+    }
+
+    this.logger.emit('mongodb-oidc-plugin:auth-succeeded', {
+      hasRefreshToken: !!state.currentTokenSet.set.refresh_token,
+      expiresAt: state.currentTokenSet.set.expires_at
+        ? new Date(state.currentTokenSet.set.expires_at * 1000).toISOString()
+        : null,
+    });
+
+    return {
+      accessToken: state.currentTokenSet.set.access_token,
+      refreshToken: state.currentTokenSet.set.refresh_token,
+      expiresInSeconds: 0,
```

Comment:
Why is this always 0?

Reply:
Yeah, great point and thanks for bringing this up. I’ve added a comment to explain this to the source code here, the tl;dr is that this disables driver caching and since we maintain the cache/state here ourselves, there’s no benefit to having a driver cache as well + the driver cache is global, not per-MongoClient, so may result in undesirable interactions between MongoClient instances.

This also reminded me to open MONGOSH-1401 to automatically schedule a timer to refresh the token before it expires, regardless of callbacks from the driver.